src/main/java/org/opentripplanner/middleware/utils/ItineraryUtils.java

```diff
@@ -93,4 +99,139 @@ public static String toQueryString(Map<String, String> params) {
+                    previousItineraryLeg.routeShortName,
+                    newItineraryLeg.routeShortName
+                ) ||
+                !equalsIgnoreCaseOrPreviousWasEmpty(
+                    previousItineraryLeg.headsign,
+                    newItineraryLeg.headsign
+                ) ||
+                !equalsOrPreviousWasNull(
+                    previousItineraryLeg.interlineWithPreviousLeg,
+                    newItineraryLeg.interlineWithPreviousLeg
+                )
+        ) {
+            return false;
+        }
+
+        // make sure the transit trips are scheduled for the same time of the day
+        if (
+            !timeOfDayMatches(
+                previousItineraryLeg.getScheduledStartTime(),
+                newItineraryLeg.getScheduledStartTime()
+            ) || !timeOfDayMatches(
+                previousItineraryLeg.getScheduledEndTime(),
+                newItineraryLeg.getScheduledEndTime()
+            )
+        ) {
+            return false;
+        }
+
+        // if this point is reached, the legs are assumed to match
+        return true;
+    }
+
+    /**
+     * Checks whether two stops (OTP Places) match for the purposes of matching itineraries
+     */
+    private static boolean stopsMatch(Place stopA, Place stopB) {
+        // stop names must match
+        if (!stopA.name.equalsIgnoreCase(stopB.name)) return false;
+
+        // stop code must match
+        if (!equalsIgnoreCaseOrPreviousWasEmpty(stopA.stopCode, stopB.stopCode)) return false;
+
+        // stop positions must be no further than 5 meters apart
+        if (
+            DistanceUtils.radians2Dist(
+                DistanceUtils.distHaversineRAD(stopA.lat, stopA.lon, stopB.lat, stopB.lon),
+                DistanceUtils.EARTH_MEAN_RADIUS_KM
+            ) * 1000 > 5
+        ) {
+            return false;
+        }
+
+        // if this point is reached, the stops are assumed to match
+        return true;
+    }
+
+    /**
+     * Returns true if previous is null. Otherwise, returns Objects.equals.
+     */
+    private static boolean equalsOrPreviousWasNull (Object previous, Object newer) {
+        return previous == null || Objects.equals(previous, newer);
+    }
+
+    /**
+     * Returns true if the previous string was empty. Otherwise, returns if the strings are equal ignoring case.
+     */
+    private static boolean equalsIgnoreCaseOrPreviousWasEmpty(String previous, String newer) {
+        return StringUtils.isEmpty(previous) || previous.equalsIgnoreCase(newer);
+    }
+
+    /**
+     * Returns true if both times have the same hour, minute and second.
+     */
+    private static boolean timeOfDayMatches(ZonedDateTime zonedDateTimeA, ZonedDateTime zonedDateTimeB) {
+        return zonedDateTimeA.getHour() == zonedDateTimeB.getHour() &&
+            zonedDateTimeA.getMinute() == zonedDateTimeB.getMinute() &&
+            zonedDateTimeA.getSecond() == zonedDateTimeB.getSecond();
```

Comment:
I find the variable names to be somewhat redundant. Definitely non-blocking, but I think omitting the `zonedDateTime` prefix would be OK here.

Reply:
When dealing with variables that represent any kind of thing having to do with time, I actually prefer to name them as explicitly as possible to avoid any confusion about what time unit they are.